src/hotspot/share/opto/addnode.cpp

```diff
@@ -1004,14 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
   const TypeInt *r0 = t0->is_int(); // Handy access
   const TypeInt *r1 = t1->is_int();
 
-  // Complementing a boolean?
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if( !r0->is_con() || !r1->is_con() ) {
+    //not a constant
 
-  if( !r0->is_con() || !r1->is_con() ) // Not constants
-    return TypeInt::INT;        // Any integer, but still no symbols.
+    // result of xor can only have bits sets where any of the
+    // inputs have bits set. lo can always become 0.
+
+    if ((r0->_lo >= 0) &&
+        (r0->_hi > 0)  &&
+        (r1->_lo >= 0) &&
+        (r1->_hi > 0)) {
```

Comment:
This can be simplified to:

    if (r0->_lo >= 0 && r1->_lo >= 0)

There is no need to check `_hi` here.

Reply:
If one arg was (lo=0,hi=1) and the other was 0 though, it might be a problem. Depends where x ^ 0 gets simplified.

Reply:
*edit2* Actually, the alternative version has an issue if either of the lo values is 0

Reply:
How about this?
    ``juint max = round_up_power_of_2<juint>( (r0->_hi | r1->_hi) +1) - 1;``

Reply:
Why? `r0->_hi | r1->_hi == 0` would imply `r0->_hi == 0 && r1->_hi == 0`. This means that both operands are constants and should be filtered out from the previous step already.

Reply:
The symptom I'm seeing with the alternative version is that the build fails with an exception of 
```
Creating jdk.net.jmod
Error: More than one SourceFile attribute
java.lang.module.InvalidModuleDescriptorException: More than one SourceFile attribute
	at java.base/jdk.internal.module.ModuleInfo.invalidModuleDescriptor(ModuleInfo.java:1228)
```
I am thinking that's a result of a bad xor. If I change the ``if ( (r0->_lo >= 0) `` to ``if ( (r0->_lo > 0)`` the build works again, but that might just be luck.

The case I'm concerned about is if `r0->_hi `and `r1->_hi` are both the same power of 2, say for example 4 and both lo values are 0. Then `round_up_power_of_2(4,4)-1` is 3. However, if the runtime value of one arg is 0 and the other is 4, `0 ^ 4 = 4, which is larger than 3.